File: Chapter02/example_publisher.rb
#	1. Require client library
require 'bunny'

#	2. Read RABBITMQ_URI from ENV
connection = Bunny.new ENV['RABBITMQ_URI']

#	3. Start a communication session with RabbitMQ
connection.start
channel = connection.create_channel



# 4. define methods
def on_start(channel)
  queue = channel.queue('taxi.1', durable: true)
  exchange = channel.direct('taxi-direct', durable: true, auto_delete: true)
  queue.bind(exchange, routing_key: 'taxi.1')
  exchange
end

def order_taxi(taxi, exchange)
  payload = "example-message"
  message_id = rand
  exchange.publish(payload,
                   routing_key: taxi,
                   content_type: "application/json",
                   content_encoding: "UTF-8",
                   persistent: true,
                   message_id: message_id)
end

#5. define exchange
exchange = on_start(channel)

#6. Publish message
order_taxi("taxi.1", exchange)

File: csharp_code/Chapter02/example_publisher.cs
using System;
using System.Collections.Generic;
using System.Text;
using RabbitMQ.Client;
using RabbitMQ.Samples.Common;

namespace RabbitMQ.Samples.Chapter02
{
    class ExamplePublisher
    {
        public static void OrderTaxi(RabbitMqConnection rabbitMqConnectionChannel)
        {
            var channel = rabbitMqConnectionChannel.Channel;

            // the actual message
            var payload = "example-message";

            // prepare the payload to be published
            var messageBodyBytes = Encoding.UTF8.GetBytes(payload);
            var props = channel.CreateBasicProperties();
            props.ContentType = "application/json";
            props.ContentEncoding = "UTF-8";
            props.Persistent = true;
            props.MessageId = Guid.NewGuid().ToString();

            // publish the message to the exchange
            channel.BasicPublish(exchange: Chapter2Constants.DirectExchange,
                routingKey: Chapter2Constants.QueueTaxi1,
                basicProperties: props, 
                body: messageBodyBytes);
        }
    }
}
